README.md
# ALSO: Allele Segregation Pipeline

Composed of a small set of Python, Shell, and R scripts, the ALSO pipeline segregates next-generation sequencing alignments to alleles of origin based on alignment scores.

## News and Updates

The three most recent updates are shown below; a complete list of updates can be found [here](https://github.com/Noble-Lab/2021_kga0_4dn-mouse-cross/blob/main/log.md).

## [0.9.8] - 2022-10-13
### changed
- changelog format from [@GangLiTarheel](https://github.com/Noble-Lab/ALSO/changelog.md).

## Installation

  + [argparser](https://bitbucket.org/djhshih/argparser) = 0.7.1
  + [bedtools](https://bedtools.readthedocs.io/en/latest/) >= 2.29.0
  + [parallel](https://www.gnu.org/software/parallel/) >= 20200101
  + [Picard](https://broadinstitute.github.io/picard/) >= 2.26.4
  + [pigz](https://zlib.net/pigz/) >= 2.3
  + [pysam](https://pypi.org/project/pysam/) >= 0.19  `# exclude-reads-from-bam.py`
  + [R](https://www.r-project.org/) >= 4.0
  + [Rsamtools](https://bioconductor.org/packages/release/bioc/html/Rsamtools.html) = 2.8.0
  + [samtools](http://www.htslib.org/) >= 1.13
  + [scales](https://scales.r-lib.org/) >= 1.1.1
  + [subread](http://subread.sourceforge.net/) = 2.0.1
  + [Tidyverse](https://www.tidyverse.org/) = 1.3.1
  + [tqdm](https://pypi.org/project/tqdm/) >= 4.64.0  `# exclude-reads-from-bam.py`

## Workflow

![plot](AlleleSegregation-05-04-2022.png)

For sci-ATAC-seq experiments part of the 4dn-mouse-cross project, the user needs to run the following steps prior to using the ALSO pipeline:
1. Demux. ([Example Code 1](https://github.com/Noble-Lab/2021_kga0_4dn-mouse-cross/blob/main/bin/workflow/01-demux.sh))
2. sci-ATAC-seq analysis pipeline from the Shendure Lab. ([Example Code 2](https://github.com/Noble-Lab/2021_kga0_4dn-mouse-cross/blob/main/bin/workflow/02-sci-ATAC-seq-analysis.sh))
3. Preprocess the bam. ([Example Code 3](https://github.com/Noble-Lab/2021_kga0_4dn-mouse-cross/blob/main/bin/workflow/03-preprocess-mm10.sh))

Example for calling `driver_allelic-segregation.sh`
```
#  Call from 2021_kga0_4dn-mouse-cross or a directory containing
#+ functions-in-progress.sh, functions-preprocessing-HPC.sh,
#+ get-AS-per-qname.R, find-set-intersection-set-complement.sh,
#+ generate-assignment-lists.R, and filter-qnames-by-assignment.sh
bash ./bin/workflow/driver_allelic-segregation.sh \
-u FALSE \
-l TRUE \
-d TRUE \
-m "512m" \
-x "4096m" \
-r "mm10" \
-s "CAST" \
-1 "path/to/mm10/Disteche_sample_N.dedup.corrected.bam" \
-2 "path/to/CAST/Disteche_sample_N.dedup.corrected.bam" \
-p "Disteche_sample_1_downsampled_test-again" \
-o "./path/to/directory/for/results" \
-b 100000 \
-c 1000000 \
-t 0 \
-a TRUE \
-n 4

# Arguments:
# -h  print this help message and exit
# -u  use safe mode: TRUE or FALSE [logical; default: FALSE]
# -l  run on GS HPC: TRUE or FALSE [logical; default: FALSE]
# -d  run pipeline in ${TMPDIR}: TRUE or FALSE [logical; default:
#     TRUE]
# -m  initial memory allocation pool for JVM [chr; default: "512m"]
# -x  maximum memory allocation pool for JVM [chr; default: "4096m"]
# -r  string for "sample #1" [chr]
# -s  string for "sample #2" [chr]
# -1  bam infile #1, including path [chr]
# -2  bam infile #2, including path [chr]
# -p  prefix for outfiles [chr]
# -o  results directory for outfiles [chr]; path will be made if it
#     does not exist
# -b  number of records to read into memory at one time when running
#     the script for Part #1, get-AS-per-qname.R [int > 0; default:
#     100000]
# -c  number of records to read into memory at one time when running
#     the script for Part #3, generate-assignment-lists.R [int > 0;
#     default: 1000000]
# -t  alignment score threshold [int >= 0; default: 0]; the absolute
#     value of the difference in alignment scores between "sample
#     #1" and "sample #2" must be greater than this value in order
#     for a sample-specific assignment to be made; if not greater than
#     this value, then the assignment will be "ambiguous"
# -a  count lines: TRUE or FALSE [logical; default: TRUE]
# -n  step in pipeline to run up to [int 1-4; default: 4]
```
 
```{bash preprocess-bam}
## 05.11
# run for all 22 samples
chmod 751 *.sh
Date="05-11"
for i in {1..22} ## run 22 samples in parallel
do
    echo "Running: "
    echo $i
    strain="mm10"
    Job_name="mm10_pre"$i
    qsub -l mfree=12G -m bea -M gangliuw@uw.edu -N $Job_name submit.sh $i ${strain} > results/${strain}-preprocessed/2022-${Date}-${strain}-pre-${i}.submit
    echo "Submitted."

    strain="CAST-EiJ"
    Job_name="CAST_pre"$i
    qsub -l mfree=12G -m bea -M gangliuw@uw.edu -N $Job_name submit.sh $i ${strain} > results/${strain}-preprocessed/2022-${Date}-${strain}-pre-${i}.submit
    echo "Submitted."
done

```

Test code for proprocessing (`workflow/03-filter-qname.sh`)
```{bash preprocess-bam-updated}
#  Call from 2021_kga0_4dn-mouse-cross
file="Disteche_sample_1.dedup.bam"
bash ./filter-qnames.sh \
-u FALSE \
-c FALSE \
-l TRUE \
-i ./"${strain}"/get_unique_fragments/"${bam_input}" \
-o "${outpath}" \
-f TRUE \
-r FALSE \
-p 4 \
> "${outpath}"preprocess_${strain}_${sample_id}.o \
2>"${outpath}"preprocess_${strain}_${sample_id}.e

# Arguments:
# -u is for "safe mode" (set -Eeuxo)
# -c is for whether using on the GS HPC or not (T or F)
# -i is for infile
# -o is for outpath
# -p is for number of cores for parallelization (for calls to samtools)
```

Example for calling `03-remove-duplicate-qnames.sh`  `#TODO` Update this...
```
#  Call from 2021_kga0_4dn-mouse-cross or a directory containing
#+ functions-in-progress.sh and functions-preprocessing-HPC.sh
bash ./bin/workflow/03-remove-duplicate-qnames.sh \
-u FALSE \
-c TRUE \
-m "512m" \
-x "4048m" \
-i "${dir_data}/${infile}" \
-o "${dir_data}" \
-n TRUE \
-t FALSE \
-e TRUE \
-r TRUE \
-p "${parallelize}" \
> "${dir_log}/rm-dup-qnames_${strain}_${ID}.o.txt" \
2> "${dir_log}/rm-dup-qnames_${strain}_${ID}.e.txt"

# ./bin/workflow/03-remove-duplicate-qnames.sh:
# Run pipeline to filter duplicate QNAMEs from bam file.
#   - Step 01: Copy files of interest to ${TMPDIR}
#   - Step 02: Sort bam by QNAME
#   - Step 03: List and tally QNAMEs in the sorted bam file
#   - Step 04: Create txt.gz outfiles for QNAME > 2
#   - Step 05: Count lines in infile, outfiles (optional)
#   - Step 06: Tally entries in infile, outfiles (optional)
#   - Step 07: Exclude problematic QNAME reads from bam infile
#   - Step 08: Sort corrected bam by QNAME (optional)
#   - Step 09: List and tally QNAMEs in the corrected bam file
#              (optional)
#   - Step 10: Create txt.gz outfiles for QNAME >, <, = 2 (optional)
#   - Step 11: Remove temporary bams, move ${TMPDIR} outfiles to
#              ${outpath}
#
#
# Dependencies:
#   - parallel >= 20200101
#   - picard >= 2.27.1
#   - samtools >= 1.13
#
#
# Arguments:
# -h print this help message and exit
# -u use safe mode: "TRUE" or "FALSE" (logical)
# -c run on GS HPC: "TRUE" or "FALSE" (logical)
# -m initial memory allocation pool for JVM (chr; default "512m")
# -x maximum memory allocation pool for JVM (chr; default "1g")
# -i bam infile, including path (chr)
# -o path for outfiles (chr); path will be made if it does not exist
# -n count lines: "TRUE" or "FALSE" (logical)
# -t tally entries: "TRUE" or "FALSE" (logical)
# -e evaluate corrected bam: "TRUE" or "FALSE" (logical)
# -r remove intermediate files: "TRUE" or "FALSE" (logical)
# -p number of cores for parallelization (int >= 1; default: 1)
```

This ALSO pipeline takes as input two paired parental bam files (one aligned to a "sample 1" reference genome and the other aligned to a "sample 2" reference genome) that have been sorted and subjected to duplicate removal; ALSO outputs 6 bam files:
  + strain-1 bam comprised of only "strain 1" assignments
  + strain-1 bam comprised of only "strain 2" assignments
  + strain-1 bam comprised of only "ambiguous" assignments
  + strain-2 bam comprised of only "strain 1" assignments
  + strain-2 bam comprised of only "strain 2" assignments
  + strain-2 bam comprised of only "ambiguous" assignments

## Proposed changes to ALSO

1. Regarding setting `-d TRUE` in various ALSO scripts: instead of copying files into and running all steps in `${TMPDIR}`, make it so that only certain programs make use of `${TMPDIR}`, e.g., Picard, Samtools; steps will be run in and files will be written to `-o "./path/to/directory/for/results"`
2. ~~Index bam outfiles while writing them out with Picard: `--CREATE_INDEX <Boolean>`~~

changelog.md

# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- main workflow figure for ALSO.
- readthedocs.


## [0.9.8] - 2022-10-13
### changed
- changelog format from [@GangLiTarheel](https://github.com/Noble-Lab/ALSO/changelog.md).

## [0.9.4] - 2022-06-25
### added
  - Developed Python implementation for filtering bam files by txt-file lists of `QNAMEs`
  - Rewrote/refactored code to handle input for Python-implemented filtering

## [0.9.3] - 2022-06-15
### fixed
  - Update `README.md` and `log.md` files
## [Unreleased]
  - `#TODO` Troubleshoot max memory for JVM when running `picard FilterSamReads`
  - `#TODO` Consolidate shell, R functions into one script for each language
  - `#TODO` Remove harcoded path to GS installation of `picard` from scripts `#DONE`

## [0.9.2] - 2022-06-04
- Pipeline is completed; passed local unit tests with small files; can call it with `driver_allelic-segregation.sh`
- Adding instructions for using `driver_allelic-segregation.sh` to `README`.
## [Unreleased]
- `#TODO` Test on the GS HPC with large files: Do we need to increase max heap memory to the JVM when running `picard`?
- `#TODO` Clean up messages output by the driver
- `#TODO` Determine and list all dependencies

## [0.9.1] - 2022-05-23
- Addressing error in preprocessing pipeline in which some duplicate QNAMEs persist in processed bam.
- Adding instructions for using the correction script, `03-remove-duplicate-qnames.sh`.
- `#TODO` Add corrections in `03-remove-duplicate-qnames.sh` to the initial preprocessing script: `03-filter-problematic-qnames-HPC.sh` `#DONE`


## [0.8.9] - 2022-05-11
### fixed
- Cleaned up old example code from [@Kris] (https://github.com/Noble-Lab/ALSO/XX).

## [Unreleased]
- Will create a pull request to move `03-remove-duplicate-qnames.sh` from the ALSO repo to the Shendure-Lab sciatac pipeline repo after the allele score comparison module is completed
- Kris will work on the allele score comparison module

## [0.8.8] - 2022-05-10
### added
- Kris' new version preprocess module passed tests from both Kris and Gang
- Gang runs the preprocess module on all samples

## [0.8.8] - 2022-05-04
### added
- Gang tested on the one sample from mm10, one sample from CAST
- Kris tested on the largest bam that we have

### fixed
- Bill cleaned up the space of vol6; all future results will be stored in vol6 
- Update the workflow according to Kris's newest preprocess module (4 steps)

## [0.8.7] - 2022-05-02
### added
- Add worflow script `03-filter-qname.sh`, an in-progress shell pipeline to handle the preprocessing
- Write code for handling intermediate files, e.g., deleting, keeping, etc.

### fixed
- Upload/update test code for debugging the preprocess module
- Update `README` for using the script
- Further updates, cleanup of the `README`

## [0.7.9] - 2022-04-13
### fixed
- Upload/update test code for debugging the preprocess module
- Update `README` for information on running the test code

## [0.7.8] - 2022-04-08
### fixed
- Replace the workflow chart
- Upload code to debug preprocess module
  
## [0.7.7] - 2022-03-27
### added
- For `04`, add additional code to remove singletons from split bam files

## [0.7.6] - 2022-03-26
### added
- Add additional options, corrections to `04-split-index-repair-bam.sh`
  - "mm10" mode, which does not output POS and MPOS bed files
  - "strain" mode, which outputs POS and MPOS bed files
  - Additional to sort and index bam infile if necessary
- Update associated test script for new modes

## [0.7.4] - 2022-03-24
### fixed
- Update workflow chart with yellow box (preprocess step)
### added
- Update run script for preprocess step

## [0.7.3] - 2022-03-23
### added
- Add `06-convert-bam-to-df_join-bed_write-rds.R`

## [0.7.2] - 2022-03-24
### removed
- Clean up repo, removing unneeded scripts and data files
### fixed
- Update dependencies listed in `README`

## [0.7.1] - 2022-03-20
### added
- Add 05-lift-strain-to-mm10.sh
- Add script to download and process liftOver chain files: `get-liftOver-chains.sh`

src_Gang/preprocess-05-10/filter-qnames.sh
#!/bin/bash

#  filter-qnames.sh
#  KA


time_start="$(date +%s)"


#  Source functions into environment ------------------------------------------
# shellcheck disable=1091
. ./functions-preprocessing-HPC.sh ||
    {
        echo "Exiting: Unable to source auxiliary information."
        exit 1
    }


#  Handle arguments, assign variables -----------------------------------------
print_usage() {
    echo ""
    echo "${0}:"
    echo "Run pipeline to filter problematic QNAMEs from bam file."
    echo "  - Step 01: Copy files of interest to \${TMPDIR}."
    echo "  - Step 02: Remove low quality reads."
    echo "  - Step 03: Sort resulting bam."
    echo "  - Step 04: Index bam."
    echo "  - Step 05: Identify, write lists of problematic QNAMEs."
    echo "  - Step 06: Collect lists in an array."
    echo "  - Step 07: Create master list of unique QNAMEs."
    echo "  - Step 08: Using master list, filter problematic QNAMEs from bam."
    echo "  - Step 09: Run samtools flagstat on bams (optional)."
    echo "  - Step 10: Sort corrected bam."
    echo "  - Step 11: Index corrected bam."
    echo "  - Step 12: Remove intermediate files from outdirectory (optional)."
    echo "  - Step 13: Move outfiles from \${TMPDIR} to \${outpath}."
    echo ""
    echo ""
    echo "Dependencies:"
    echo " - argparser >= 0.7.1"
    echo " - picard >= 2.27.1"
    echo " - R >= 4.0.5"
    echo " - Rsamtools >= 2.6.0"
    echo " - Rscript >= 4.0.5"
    echo " - samtools >= 1.13"
    echo " - scales >= 1.1.1"
    echo " - tidyverse >= 1.3.1"
    echo ""
    echo ""
    echo "Arguments:"
    echo "-h print this help message and exit"
    echo "-u use safe mode: \"TRUE\" or \"FALSE\" (logical)"
    echo "-c use KA conda environment: \"TRUE\" or \"FALSE\" (logical)"
    echo "-l run on GS HPC: \"TRUE\" or \"FALSE\" (logical)"
    echo "-i bam infile, including path (chr)"
    echo "-o path for outfiles (chr); path will be made if it does not exist"
    echo "-f run samtools flagstat on bams: \"TRUE\" or \"FALSE\" (logical)"
    echo "-r remove intermediate files: \"TRUE\" or \"FALSE\" (logical)"
    echo "-p number of cores for parallelization (int >= 1); default: 1"
    exit
}


while getopts "h:u:c:l:i:o:f:r:p:" opt; do
    case "${opt}" in
        h) print_usage ;;
        u) safe_mode="${OPTARG}" ;;
        c) conda="${OPTARG}" ;;
        l) cluster="${OPTARG}" ;;
        i) infile="${OPTARG}" ;;
        o) outpath="${OPTARG}" ;;
        f) flagstat="${OPTARG}" ;;
        r) remove="${OPTARG}" ;;
        p) parallelize="${OPTARG}" ;;
        *) print_usage ;;
    esac
done

[[ -z "${safe_mode}" ]] && print_usage
[[ -z "${conda}" ]] && print_usage
[[ -z "${cluster}" ]] && print_usage
[[ -z "${infile}" ]] && print_usage
[[ -z "${outpath}" ]] && print_usage
[[ -z "${flagstat}" ]] && print_usage
[[ -z "${remove}" ]] && print_usage
[[ -z "${parallelize}" ]] && parallelize=1


#  Check variable assignments -------------------------------------------------
echo -e ""
echo -e "Running ${0}... "

#  Evaluate "${safe_mode}"
case "$(echo "${safe_mode}" | tr '[:upper:]' '[:lower:]')" in
    true | t) echo -e "-u: Safe mode is on." && set -Eeuxo pipefail ;;
    false | f) echo -e "-u: Safe mode is off." ;;
    *) \
        echo -e "Exiting: -u safe mode argument must be \"TRUE\" or \"FALSE\".\n"
        exit 1
        ;;
esac

#  Evaluate "${conda}"
case "$(echo "${conda}" | tr '[:upper:]' '[:lower:]')" in
    false | f) \
        echo -e "-c: Use KA conda environment is FALSE."
        ;;
    true | t) \
        #  Conda anvironment used by KA for writing and testing the pipeline
        echo -e "-c: Use KA conda environment is TRUE."
        conda activate pipeline-test_env  
        ;;
    *) \
        echo -e "Exiting: -c argument must be \"TRUE\" or \"FALSE\".\n"
        return 1
        ;;
esac

#  Check for necessary dependencies; exit if not found
check_dependency R
check_dependency samtools

#  Evaluate "${cluster}"
case "$(echo "${cluster}" | tr '[:upper:]' '[:lower:]')" in
    true | t) echo -e "-l: Run on GS HPC is TRUE." ;;
    false | f) echo -e "-l: Run on GS HPC is FALSE." && check_dependency picard ;;
    *) \
        echo -e "Exiting: -l argument must be \"TRUE\" or \"FALSE\".\n"
        return 1
        ;;
esac

#  Check that "${infile}" exists
[[ -f "${infile}" ]] ||
    {
        echo -e "Exiting: -i ${infile} does not exist.\n"
        exit 1
    }

#  Make "${outpath}" if it doesn't exist
[[ -d "${outpath}" ]] ||
    {
        echo -e "-o: Directory ${outpath} does not exist; making the directory."
        mkdir -p "${outpath}"
    }

#  Evaluate "${flagstat}"
case "$(echo "${flagstat}" | tr '[:upper:]' '[:lower:]')" in
    true | t) \
        flagstat=1
        echo -e "-f: Run samtools flagstat is TRUE."
        ;;
    false | f) \
        flagstat=0
        echo -e "-f: Run samtools flagstat is FALSE."
        ;;
    *) \
        echo -e "Exiting: -f flagstat argument must be \"TRUE\" or \"FALSE\".\n"
        exit 1
        ;;
esac

#  Evaluate "${remove}"
case "$(echo "${remove}" | tr '[:upper:]' '[:lower:]')" in
    true | t) \
        remove=1
        echo -e "-r: Remove intermediate files is TRUE."
        ;;
    false | f) \
        remove=0
        echo -e "-r: Remove intermediate files is FALSE."
        ;;
    *) \
        echo -e "Exiting: -f flagstat argument must be \"TRUE\" or \"FALSE\".\n"
        exit 1
        ;;
esac

#  Check "${parallelize}"
[[ ! "${parallelize}" =~ ^[0-9]+$ ]] &&
    {
        echo -e "Exiting: -p parallelize argument must be an integer.\n"
        exit 1
    }

[[ ! $((parallelize)) -ge 1 ]] &&
    {
        echo -e "Exiting: -p parallelize argument must be an integer >= 1.\n"
        exit 1
    }

echo ""


#  Assign variables needed for the pipeline -------------------------------
base="$(basename "${infile}")"
base_tmp="${TMPDIR}/${base}"
base_rm="${base%.bam}.rm.bam"
base_rm_bai="${base%.bam}.rm.bam.bai"
out_rm="${TMPDIR}/${base_rm}"
out_rm_bai="${TMPDIR}/${base_rm_bai}"
out_ambiguous="${base%.bam}.rm.ambiguous.txt.gz"
out_duplicated="${base%.bam}.rm.duplicated.txt.gz"
out_singleton="${base%.bam}.rm.singleton.txt.gz"
out_trans="${base%.bam}.rm.trans.txt.gz"
out_unmated="${base%.bam}.rm.unmated.txt.gz"
out_exclude="${TMPDIR}/${base%.bam}.to-exclude.txt.gz"
out_corrected="${TMPDIR}/${base%.bam}.corrected.bam"
out_flagstat_infile="${TMPDIR}/${base%.bam}.flagstat.txt"
out_flagstat_rm="${TMPDIR}/${base%.bam}.rm.flagstat.txt"
out_flagstat_corrected="${TMPDIR}/${base%.bam}.corrected.flagstat.txt"

#  Assign variables for completion files
step_1="$(echo_completion_file "${outpath}" 1 "${infile}")"
step_2="$(echo_completion_file "${outpath}" 2 "${infile}")"
step_3="$(echo_completion_file "${outpath}" 3 "${infile}")"
step_4="$(echo_completion_file "${outpath}" 4 "${infile}")"
step_5="$(echo_completion_file "${outpath}" 5 "${infile}")"
step_6="$(echo_completion_file "${outpath}" 6 "${infile}")"
step_7="$(echo_completion_file "${outpath}" 7 "${infile}")"
step_8="$(echo_completion_file "${outpath}" 8 "${infile}")"
step_9="$(echo_completion_file "${outpath}" 9 "${infile}")"
step_10="$(echo_completion_file "${outpath}" 10 "${infile}")"
step_11="$(echo_completion_file "${outpath}" 11 "${infile}")"
step_12="$(echo_completion_file "${outpath}" 12 "${infile}")"
step_13="$(echo_completion_file "${outpath}" 13 "${infile}")"


#  01: Copy files of interest to ${TMPDIR} ------------------------------------
if [[ ! -f "${step_1}" ]]; then
    echo -e "Started step 1/13: Copying ${base} into ${TMPDIR}."
    cp "${infile}" "${TMPDIR}" && \
    touch "${step_1}" && \
    echo -e "Completed step 1/13: Copying ${base} into ${TMPDIR}.\n"
else
    echo_completion_message 1
    :
fi


#  02: Remove low-quality reads (-f 3 -F 12 -q 30) ----------------------------
if [[ ! -f "${step_2}" && -f "${step_1}" ]]; then
    echo -e "Started step 2/13: Running samtools view (-f 3 -F 12 -q 30) on ${base_tmp}."
    remove_reads_low_quality "${parallelize}" "${base_tmp}" "${out_rm}" && \
    touch "${step_2}" && \
    echo -e "Completed step 2/13: Running samtools view (-f 3 -F 12 -q 30) on ${base_tmp}.\n"
elif [[ -f "${step_2}" && -f "${step_1}" ]]; then
    echo_completion_message 2
    :
else
    echo_exit_message 2
    exit 1
fi



#  03: Sort bam by coordinate, writing a tmp file that overwrites the infile --
if [[ ! -f "${step_3}" && -f "${step_2}" ]]; then
    echo -e "Started step 3/13: Running samtools sort on ${out_rm}."
    sort_bam_coordinate_samtools_overwrite_infile \
    "${parallelize}" "${out_rm}" && \
    touch "${step_3}" && \
    echo -e "Completed step 3/13: Running samtools sort on ${out_rm}.\n"
elif [[ -f "${step_3}" && -f "${step_2}" ]]; then
    echo_completion_message 3
    :
else
    echo_exit_message 3
    exit 1
fi


#  04: Index bam file ---------------------------------------------------------
if [[ ! -f "${step_4}" && -f "${step_3}" ]]; then
    echo -e "Started step 4/13: Running samtools index on ${out_rm}."
    index_bam "${parallelize}" "${out_rm}" && \
    touch "${step_4}" && \
    echo -e "Completed step 4/13: Running samtools index on ${out_rm}.\n"
elif [[ -f "${step_4}" && -f "${step_3}" ]]; then
    echo_completion_message 4
    :
else
    echo_exit_message 4
    exit 1
fi


#  05: Generate lists of QNAMEs to exclude ------------------------------------
if [[ ! -f "${step_5}" && -f "${step_4}" ]]; then
    echo -e "Started step 5/13: Generating lists of QNAMEs to exclude from ${out_rm}."
    Rscript ./generate-qname-lists.R \
    --bam "${out_rm}" \
    --bai "${out_rm_bai}" \
    --outdir "${TMPDIR}" \
    --chunk 100000 \
    --mated FALSE \
    --unmated TRUE \
    --ambiguous TRUE \
    --trans TRUE \
    --duplicated TRUE \
    --singleton TRUE \
    --unique TRUE \
    --tally FALSE \
    --remove TRUE
    
    if [[ -f "${TMPDIR}/${out_unmated}" ]]; then
        touch "${step_5}"
    else
        echo -e "There was a problem: ${TMPDIR}/${out_unmated} was not generated."
        echo -e "Check on this."
        echo_exit_message 6
        exit 1
    fi
    echo -e "Completed step 5/13: Generating lists of QNAMEs to exclude from ${out_rm}.\n"
elif [[ -f "${step_5}" && -f "${step_4}" ]]; then
    echo_completion_message 5
    :
else
    echo_exit_message 5
    exit 1
fi


#  06: Collect outfiles in an array -----------------------------------------
find_outfiles() {
    #TODO Description of function
    #
    # :param TMPDIR: path for outfiles (chr)
    # :param out_ambiguous: txt.gz list of ambiguous QNAMEs (chr) 
    # :param out_duplicated: txt.gz list of duplicated QNAMEs (chr) 
    # :param out_singleton: txt.gz list of singleton QNAMEs (chr) 
    # :param out_trans: txt.gz list of trans QNAMEs (chr) 
    # :param out_unmated: txt.gz list of unmated QNAMEs (chr) 
    find "${TMPDIR}" \
    -maxdepth 1 \
    -type f \
    \( -name "${out_ambiguous}" -o \
    -name "${out_duplicated}" -o \
    -name "${out_singleton}" -o \
    -name "${out_trans}" -o \
    -name "${out_unmated}" \) \
    -print0
}


if [[ ! -f "${step_6}" && -f "${step_5}" ]]; then
    echo -e "Started step 6/13: Collecting outfiles into an array."
    unset outfiles
    typeset -a outfiles
    while IFS=" " read -r -d $'\0'; do
        outfiles+=( "${REPLY}" )
    done < <(find_outfiles)

    if [[ ${#outfiles[@]} -eq 0 ]]; then
        echo -e "There was a problem: The number of outfiles is ${#outfiles[@]}."
        echo -e "Check on this."
        echo_exit_message 6
        exit 1
    else
        echo -e "Number of elements in \${outfiles[@]}: ${#outfiles[@]}"
        echo -e "Elements in \${outfiles[@]}:"
        echo_loop "${outfiles[@]}"; echo -e ""
        touch "${step_6}" && \
        echo -e "Completed step 6/13: Collecting outfiles into an array.\n"
    fi
elif [[ -f "${step_6}" && -f "${step_5}" ]]; then
    unset outfiles
    typeset -a outfiles
    while IFS=" " read -r -d $'\0'; do
        outfiles+=( "${REPLY}" )
    done < <(find_outfiles)

    if [[ ${#outfiles[@]} -eq 0 ]]; then
        echo -e "There was a problem: The number of outfiles is ${#outfiles[@]}."
        echo -e "Check on this."
        echo_exit_message 6
        exit 1
    else
        echo -e "Number of elements in \${outfiles[@]}: ${#outfiles[@]}"
        echo -e "Elements in \${outfiles[@]}:"
        echo_loop "${outfiles[@]}"; echo -e ""
    fi

    echo_completion_message 6
    :
else
    echo_exit_message 6
    exit 1
fi


#  07: Combine outfiles into one file for excluding problematic QNAME reads ---
if [[ ! -f "${step_7}" && -f "${step_6}" ]]; then
    echo -e "Started step 7/13: Combining outfiles into one file, $(basename "${out_exclude}")."
    combine_gz_qname_lists_return_unique_gzip "${outfiles[@]}" \
    > "${out_exclude}" && \
    touch "${step_7}"

    records=$(count_lines_gzip "${out_exclude}" 2> /dev/null)
    if [[ $(( records )) -eq 0 ]]; then
        echo -e "There was a problem generating $(basename "${out_exclude}"), which"
        echo -e "currently contains $(( records )) lines. Check on this."
        echo_exit_message 7
    else
        echo -e "$(basename "${out_exclude}"): $(( records )) lines"
        echo -e "Completed step 7/13: Combining outfiles into one file, $(basename "${out_exclude}").\n"
    fi
elif [[ -f "${step_7}" && -f "${step_6}" ]]; then
    if [[ -f "${out_exclude}" ]]; then
        records=$(count_lines_gzip "${out_exclude}" 2> /dev/null)
        if [[ $(( records )) -eq 0 ]]; then
            echo -e "There was a problem generating $(basename "${out_exclude}"), which"
            echo -e "currently contains $(( records )) lines. Check on this."
            echo_exit_message 7
        else
            echo -e "$(basename "${out_exclude}"): $(( records )) lines"
            echo_completion_message 7
            :
        fi
    else
        echo "${out_exclude} is missing. Check on this."
        echo_exit_message 7
    fi
else
    echo_exit_message 7
    exit 1
fi


#  08: Exclude problematic QNAME reads from bam infile ------------------------
if [[ ! -f "${step_8}" && -f "${step_7}" ]]; then
    echo -e "Started step 8/13: Using ${out_exclude} to exclude reads from ${out_rm}."
    exclude_qname_reads_picard \
    "${out_rm}" \
    "${out_exclude}" \
    "${out_corrected}" \
    "${cluster}" && \
    touch "${step_8}" && \
    echo -e "Completed step 8/13: Using ${out_exclude} to exclude reads from ${out_rm}.\n"
elif [[ -f "${step_8}" && -f "${step_7}" ]]; then
    echo_completion_message 8
    :
else
    echo_exit_message 8
    exit 1
fi


#  09: Run flagstat on bam in and outfiles (optional) -------------------------
if [[ ! -f "${step_9}" && -f "${step_8}" ]]; then
    echo -e "Started step 9/13 (optional): Running samtools flagstat on ${infile}, ${out_rm}, and ${out_corrected}."
    if [[ "${flagstat}" == 1 ]]; then
        run_flagstat "${parallelize}" "${infile}" "${out_flagstat_infile}"
        run_flagstat "${parallelize}" "${out_rm}" "${out_flagstat_rm}"
        run_flagstat "${parallelize}" "${out_corrected}" "${out_flagstat_corrected}"

        echo_flagstat "${out_flagstat_infile}"
        echo_flagstat "${out_flagstat_rm}"
        echo_flagstat "${out_flagstat_corrected}"
        echo -e ""
    fi
    touch "${step_9}" && \
    echo -e "Completed step 9/13 (optional): Running samtools flagstat on ${infile}, ${out_rm}, and ${out_corrected}.\n"
elif [[ -f "${step_9}" && -f "${step_8}" ]]; then
    if [[ "${flagstat}" == 1 ]]; then
        echo_completion_message 9
        echo_flagstat "${out_flagstat_infile}" || echo -e "${out_flagstat_infile} not found."
        echo_flagstat "${out_flagstat_rm}" || echo -e "${out_flagstat_rm} not found."
        echo_flagstat "${out_flagstat_corrected}" || echo -e "${out_flagstat_corrected} not found."
        echo -e ""
    fi
else
    echo_exit_message 9
    exit 1
fi


#  10: Sort corrected bam -----------------------------------------------------
if [[ ! -f "${step_10}" && -f "${step_9}" ]]; then
    echo -e "Started step 10/13: Running samtools sort on ${out_corrected}."
    sort_bam_coordinate_samtools_overwrite_infile \
    "${parallelize}" "${out_corrected}" && \
    touch "${step_10}" && \
    echo -e "Completed step 10/13: Running samtools sort on ${out_corrected}.\n"
elif [[ -f "${step_10}" && -f "${step_9}" ]]; then
    echo_completion_message 10
    :
else
    echo_exit_message 10
    exit 1
fi


#  11: Index corrected bam ----------------------------------------------------
if [[ ! -f "${step_11}" && -f "${step_10}" ]]; then
    echo -e "Started step 11/13: Running samtools index on ${out_corrected}."
    index_bam "${parallelize}" "${out_corrected}" && \
    touch "${step_11}" && \
    echo -e "Completed step 11/13: Running samtools index on ${out_corrected}.\n"
elif [[ -f "${step_11}" && -f "${step_10}" ]]; then
    echo_completion_message 11
    :
else
    echo_exit_message 11
    exit 1
fi


#  12: Remove unneeded intermediate files (optional) ---------------------------
rm -f "${base_tmp}"  # Remove infile copied into "${TMPDIR}"
if [[ ! -f "${step_12}" && -f "${step_11}" ]]; then
    echo -e "Started step 12/13 (optional): Removing unneeded intermediate files."
    if [[ "${remove}" == 1 ]]; then
        rm -f "${outfiles[@]}" "${out_rm}" "${out_rm_bai}"
    fi
    touch "${step_12}" && \
    echo -e "Completed step 12/13 (optional): Removing unneeded intermediate files.\n"
elif [[ -f "${step_12}" && -f "${step_11}" ]]; then
    echo_completion_message 12
    :
else
    echo_exit_message 12
    exit 1
fi


#  13: Move outfiles from "${TMPDIR}" to "${outpath}" -------------------------
if [[ -f "${step_12}" && -f "${out_corrected}" ]]; then
    echo -e "Started step 13/13: Moving outfiles from ${TMPDIR} to ${outpath}."
    mv -f "${TMPDIR}/"*.{bam,bam.bai,txt,txt.gz} "${outpath}" && \
    touch "${step_13}" && \
    echo -e "Completed step 13/13: Moving outfiles from ${TMPDIR} to ${outpath}.\n"
fi


#  Return run time ------------------------------------------------------------
time_end="$(date +%s)"
calculate_run_time "${time_start}" "${time_end}" "Completed: ${0}"
echo -e ""

exit 0